### PR TITLE
Generic signature fixes

### LINKED ISSUE #10837: Needless (incorrect?) Signature added to Java varargs methods
## Minimized code

```Scala
import scala.annotation.varargs

object A {
  @varargs
  def create(args: String*): String = null
}
class A
```

## Output

```
$ javap -v A
  public static java.lang.String create(java.lang.String...);
    descriptor: ([Ljava/lang/String;)Ljava/lang/String;
    flags: (0x0089) ACC_PUBLIC, ACC_STATIC, ACC_VARARGS
[..]
    Signature: #17                          // ([Ljava/lang/String;)Ljava/lang/String;
```

## Expectation

No Signature, as it matches the descriptor (reified arrays).

Came up testing MiMa with Scala 3.0.0-M3, but I consider it a niche issue in the Scala 3 compiler.

### PR DESCRIPTION
**Pull request by SolalPirelli**
Fixes #10837

* Only emit generic signatures when they differ from the descriptor
* Remove the convoluted logic for trying to guess whether a signature is needed, instead:
  * after generation, we check if the signature is equal to the descriptor and ignore it if it is
  * before generation, we check some extremely simple cases for which we know signatures aren't needed
* Fix a number of edge cases, see test changes
* Micro-optimizations to improve performance, such as iterating directly over list buffers instead of copying them to lists, not creating intermediate lists, etc.
  * Overall perf (see last comment) is +/- the same as before

## How much have you relied on LLM-based tools in this contribution?

Not at all

## How was the solution tested?

Automated tests

### CODE REVIEW

File: compiler/src/dotty/tools/backend/jvm/BCodeHelpers.scala [hunk 1]
```diff
@@ -686,32 +686,21 @@ trait BCodeHelpers(val backendUtils: BackendUtils)(using ctx: Context) extends B
...
-    if (erasedMemberType =:= sym.denot.info)
-      getGenericSignatureHelper(sym, moduleClass, memberTpe).orNull
+    if !ctx.base.settings.XnoGenericSig.value then
+      val memberTpe = atPhase(erasurePhase) { moduleClass.denot.thisType.memberInfo(sym) }
+      val erasedMemberType = ElimErasedValueType.elimEVT(TypeErasure.transformInfo(sym, memberTpe))
+      if (erasedMemberType =:= sym.denot.info)
+        val gensig = getGenericSignatureHelper(sym, moduleClass, memberTpe).orNull
+        if gensig == descriptor then null
+        else gensig
+      else null
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/backend/jvm/BCodeHelpers.scala:703]: Not pretty, but this method belongs in the lowest circle of Hell anyway and should disappear if I ever have the courage to clean up mirror generic signature generation.

**Comment by lrytz** [comment]
In Scala 2 this is handled by `needsJavaSig` returning false. Probably that should also be false in Scala 3 for the varargs forwarder?

**Comment by SolalPirelli** [comment]
I could've added another special case but I didn't get why we even have this "needs generic signature" test, there are probably other such cases of methods that shouldn't have a generic signature but do right now, it's hard to test, and all tests pass with this change.

Is there some drawback to the "always generate unless it's been turned off or it's unneeded" approach that I'm not aware of?

**Comment by lrytz** [comment]
> I could've added another special case

But it shouldn't need a special case; `needsJavaSig` should be false for `(Array[String]): String`

> I didn't get why we even have this "needs generic signature" test

I don't know either... maybe git history would reveal something.

From the top of my head what you say makes sense, we don't need a signature that duplicates the descriptor. But it feels a bit like patching things up given there's the `needsJavaSig` check already.

**Comment by SolalPirelli** [comment]
> But it shouldn't need a special case; needsJavaSig should be false for (Array[String]): String

The reason it's currently true isn't the type itself but the fact that method is considered a bridge, and for some reason we generate generic signatures for all such methods. (I think it might also have the artifact flag... I don't remember exactly since I looked at it)

As to why we check these flags, it dates back to the introduction of GenASM (i.e., it was in the commit that added that file), so I guess we'll never know 😅

**Comment by tanishiking** [comment]
>  I don't know how to check for the existence of a signature in a non-generic method programmatically without importing ASM in the test which seems too heavyweight.

I found that there's `DottyBytecodeTests` that we can access to ASM, and there're some tests checking it's signatures. I guess we can test this by `.signature == null` ?

edit: oh, you've already tweaked the test in this PR

https://github.com/scala/scala3/blob/88140bc8bfe3ac77540f5e092d06caf2866ffc76/compiler/test/dotty/tools/backend/jvm/DottyBytecodeTests.scala#L1715-L1734

**Comment by SolalPirelli** [comment]
True, I don't like the fact we have this separate way of testing things but it's there so we can use it... added a test

**Comment by lrytz** [comment]
> > But it shouldn't need a special case; needsJavaSig should be false for (Array[String]): String
> 
> The reason it's currently true isn't the type itself but the fact that method is considered a bridge, and for some reason we generate generic signatures for all such methods.

I cannot spot this in `needsJavaSig` - where is it? I think the right place to fix this issue would be there.

**Comment by SolalPirelli** [comment]
ok sorry it's the other way round:

- we check if the varargs is a Java override here: https://github.com/scala/scala3/blob/main/compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala#L69-L70
- it's not in this simple case, so we don't set `Artifact` here: https://github.com/scala/scala3/blob/main/compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala#L250
- as a result, all three of these conditions are false -- note that the method wraps them in an `!(...)`: https://github.com/scala/scala3/blob/main/compiler/src/dotty/tools/backend/jvm/BCodeHelpers.scala#L697-L699

We can't simply declare "if `JavaVarargs`, then no generic signature" since those are orthogonal concepts, and I'm not really keen on adding flags like `Artifact` because those tend to have widespread ramifications...

Also, no clue how many other such special cases exist. I think "always generate generic signatures if the descriptor isn't enough" simplifies the code overall. The output could in theory be larger, but since right now we generate unneeded signatures as well, this isn't the case: the stdlib is `9640782` bytes on main and `9626129` bytes on this PR (rebased on main).

**Comment by lrytz** [comment]
Ah, so there's `needsGenericSignature` in the backend, but also `GenericSignatures.needsJavaSig`. I was talking about the latter, I think it should return false on the type `(Array[String]): String`.

**Comment by SolalPirelli** [comment]
OK, that makes sense, then should we have an assert that we never generate a generic signature that is equal to the descriptor? (Presumably there are other missing cases like this, but those can be fixed)

**Comment by tanishiking** [comment]
>  then should we have an assert that we never generate a generic signature that is equal to the descriptor?

If the issue can be fixed in `GenericSignatures.needsJavaSig`, I personally don't think we don't need an assertion in the downstream (if unnecessary generic signature isn't a critical problem).

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/25665787405).

**Comment by mbovel** [comment]
Benchmarks completed. [Overview](https://lampepfl.github.io/scala3-benchmarks/#compare/3.9.0-RC1-bin-20260511-c3e3e42-NIGHTLY,3.9.0-RC1-bin-e6df87f428b507f773ba3b6384f461e92a2a3ed9-BENCH).

**Comment by SolalPirelli** [comment]
Interesting results from the benchmarks above, seems there's a small but real regression on some of the benchmarks, though also small improvements on others.

That's with changes where I simply removed the check for whether we need a generic signature from `GenericSignatures`, instead always generating it and then checking whether it's equal to the descriptor.

One advantage of this approach is that we don't have duplicated logic, and another is that we will always check the signatures if the flag to check them is passed, which is the case for tests.

As a result, I noticed `main` doesn't handle `@targetName` in generic signatures. There are no generic-signature-specific tests with it, but we have a test that uses it, and with this simpler logic the signature was emitted since it differed from the descriptor and then `javac` caught the problem. I'll add a generic-signature-specific test.

I wonder if I can get this to acceptable perf.

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/25794684738).

**Comment by mbovel** [comment]
Benchmarks failed. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/25794684738).

**Comment by mbovel** [comment]
Re-trying with 3 runs.

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/25799742006).

**Comment by mbovel** [comment]
Benchmarks completed. [Overview](https://lampepfl.github.io/scala3-benchmarks/#compare/3.9.0-RC1-bin-20260511-c3e3e42-NIGHTLY,3.9.0-RC1-bin-e6df87f428b507f773ba3b6384f461e92a2a3ed9-BENCH).

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/25908789735).

**Comment by mbovel** [comment]
Benchmarks completed. [Overview](https://lampepfl.github.io/scala3-benchmarks/#compare/3.9.0-RC1-bin-20260515-ed18d94-NIGHTLY,3.9.0-RC1-bin-e01e3123222d9f682227857ff99839012e240291-BENCH).

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/25963253317).

**Comment by mbovel** [comment]
Benchmarks completed. [Overview](https://lampepfl.github.io/scala3-benchmarks/#compare/3.9.0-RC1-bin-20260515-ed18d94-NIGHTLY,3.9.0-RC1-bin-4f859d678f61478656c481eced282eb7ebcc98de-BENCH).

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/26025606427).

**Comment by mbovel** [comment]
Benchmarks completed. [Overview](https://lampepfl.github.io/scala3-benchmarks/#compare/3.9.0-RC1-bin-20260518-0ce4c47-NIGHTLY,3.9.0-RC1-bin-0f57263f2e6eea30115799ce5ef9d7e708916643-BENCH).

**Comment by mbovel** [comment]
Benchmarks started. [Workflow run](https://github.com/lampepfl/scala3-benchmarks/actions/runs/26089942619).

**Comment by mbovel** [comment]
Benchmarks completed. [Overview](https://lampepfl.github.io/scala3-benchmarks/#compare/3.9.0-RC1-bin-0c45af935304406f4f435ab8059282ac9efe7a2f-BENCH,3.9.0-RC1-bin-5e02c997aab16e6df774cc6b22de5d7e0a856bdc-BENCH).

File: tests/run/targetName-modules-1/Test_2.java [hunk 2]
```diff
@@ -4,6 +4,6 @@ public class Test_2 {
 
   public static void main(String[] args) {
     assert Outer$.A.foo() == 23;
-    assert Outer$A$.MODULE$.foo() == 23;
+    assert Outer.A$.MODULE$.foo() == 23;
```
Comment (SolalPirelli) [inline tests/run/targetName-modules-1/Test_2.java:7]: Corresponding Scala def in the same folder, for reference:
```scala
package alpha

object Outer {
  @scala.annotation.targetName("A") object B {
    def foo = 23
  }
}
```

I don't think it makes sense that one could access this as `Outer$A$`, it's a nested class, we have tests for this behavior (`Outer.A$`) without `@targetName` already and a comment in `BTypes.scala` that specifically says this should be accessed as `Outer.Inner` from Java.

File: tests/generic-java-signatures/targetName.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import scala.annotation.targetName
```
Comment (SolalPirelli) [inline tests/generic-java-signatures/targetName.scala:1]: This test was partially failures (not using the `targetName` in the signature) and partially a crash (for `XO`), both now fixed.

File: compiler/src/dotty/tools/backend/jvm/BTypeLoader.scala [hunk 4]
```diff
@@ -294,7 +294,7 @@ final class BTypeLoader(primitives: ScalaPrimitives, inlineInfoLoader: () => Opt
         if (innerClassSym.isAnonymousClass || innerClassSym.isAnonymousFunction) None
         else {
           val original = innerClassSym.initial
-          Some(atPhase(original.validFor.lastPhaseId)(innerClassSym.name).mangledString) // moduleSuffix for module classes
+          Some(atPhase(original.validFor.lastPhaseId)(innerClassSym.targetName).mangledString) // moduleSuffix for module classes
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/backend/jvm/BTypeLoader.scala:297]: Without this, we emit the wrong name for inner class attributes, so the newly-added test fails. Interestingly, it seems that when loading a generic signature Java can overwrite the name with the inner class name internally, so even after I fixed `GenericSignatures.java` to use `targetName` the test still failed even though the emitted signature (at the level of method signatures) was correct.

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 5]
```diff
@@ -190,16 +192,6 @@ object GenericSignatures {
...
 
-    // Anything which could conceivably be a module (i.e. isn't known to be
-    // a type parameter or similar) must go through here or the signature is
-    // likely to end up with Foo<T>.Empty where it needs Foo<T>.Empty$.
-    def fullNameInSig(sym: Symbol): Unit = {
-      assert(sym.isClass)
-      // Time travel necessary so we get the full name after inner classes have been lifted to package scope
-      val name = atPhase(flattenPhase.next) { sanitizeName(sym.fullName).replace('.', '/') }
-      builder.append('L').append(name)
-    }
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:201]: Inlined into its only use below

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 6]
```diff
@@ -239,33 +231,10 @@ object GenericSignatures {
               // Hence the widenNullaryMethod.
         }
 
-      if (pre.exists) {
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:242]: This wasn't necessary, the `fullNameInSig` logic already did that. In fact this logic was slightly wrong for some cases (e.g., using `.` only after `>` is not quite right, inner classes can be in non-generic outer classes)

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 7]
```diff
@@ -278,7 +247,17 @@ object GenericSignatures {
...
+        if (isGenericArrayElement(elemtp, isScala2 = false))
+          jsig(defn.ObjectType)
+        else
+          builder.append(ClassfileConstants.ARRAY_TAG)
+          elemtp match
+            case TypeBounds(lo, hi) => jsig(hi.widenDealias)
+            // derived VCs are not unboxed inside arrays,
+            // i.e., `Array[VC]` where `class VC(n: X) extends AnyVal`
+            // is `[LVC;`, not `[LX;`
+            case _ => jsig(elemtp, vcBoxing = ValueClassBoxing.UnboxOnlyPrimitives)
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:271]: Existing code extracted to a method so we also handle Java array types

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 8]
```diff
@@ -313,99 +291,95 @@ object GenericSignatures {
...
-          else if sym.isClass then
-            classSig(sym, pre, args)
-          else
-            jsig(erasure(tp), toplevel = toplevel, vcBoxing = vcBoxing)
-
-        case ExprType(restpe) if toplevel =>
-          builder.append("()")
-          methodResultSig(restpe)
+            typeParamSig(sym.targetName.lastPart)
+          else defn.specialErasure.get(sym) match
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:314]: Review ignoring whitespace, had to do this change to not do two lookups in `defn.specialErasure`

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 9]
```diff
@@ -433,51 +407,41 @@ object GenericSignatures {
...
           jsig(hktl.finalResultType, toplevel, vcBoxing)
 
+        case ErasedValueType(tycon, underlying) =>
+          if vcBoxing == ValueClassBoxing.Unbox || (vcBoxing == ValueClassBoxing.UnboxOnlyPrimitives && underlying.isPrimitiveValueType)
+          then jsig(underlying, toplevel, vcBoxing)
+          else jsig(tycon, toplevel, vcBoxing)
+
         case _ =>
-          val etp = erasure(tp)
-          if (etp eq tp) throw new UnknownSig
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:438]: This happening is a bug. It doesn't happen anymore. So we no longer need exceptions for control flow thankfully.

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 10]
```diff
@@ -48,7 +48,8 @@ class TreeChecker extends Phase with SymTransformer {
   def testDuplicate(sym: Symbol, registry: mutable.Map[String, Symbol], typ: String)(using Context): Unit = {
     val name = sym.javaClassName
     val isDuplicate = this.flatClasses && registry.contains(name)
-    assert(!isDuplicate, s"$typ defined twice $sym ${sym.id} ${registry(name).id}")
+    // Allow users to define a class "java" even on the JVM
+    assert(!isDuplicate || name == "java", s"$typ defined twice $sym ${sym.id} ${registry(name).id}")
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:52]: We have one test that's a file with just `class java`. It already bit me with optimizer work, because it means you can get into `-Ycheck` failures if you, e.g., load `java.lang.Object` eagerly.

File: tests/generic-java-signatures/derivedNames.scala [hunk 11]
```diff
@@ -4,12 +4,12 @@ object Test {
   def main(args: Array[String]): Unit = {
     val objectB = classOf[Foo[Any]].getClasses
     val returnType = objectB(1).getDeclaredMethod("m").getGenericReturnType.asInstanceOf[ParameterizedType]
-    val out1 = "Test$Foo.Test$Foo$A<Test.Test$Foo<T1>.B$>" // Windows
-    val out2 = "Test$Foo$A<Test$Foo<T1>$B$>"               // Linux, OSX and sometimes Windows
+    val out1 = "Test$Foo.Test$Foo$A<Test.Test$Foo.B$>" // Windows
+    val out2 = "Test$Foo$A<Test$Foo$B$>"               // Linux, OSX and sometimes Windows
```
Comment (SolalPirelli) [inline tests/generic-java-signatures/derivedNames.scala]: This was just wrong? There's no reason to include a generic parameter like this in a signature, AFAIK?

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 12]
```diff
@@ -32,21 +29,29 @@ object GenericSignatures {
...
-    // names which are not valid Java identifiers (see git history of this method).
+  def javaSig(sym0: Symbol, info: Type)(using Context): StringBuilder | Null =
+    if mayNeedSignature(sym0, info) then atPhase(erasurePhase)(javaSig0(sym0, info))
+    else null
+
+  private def mayNeedSignature(sym0: Symbol, info: Type)(using Context) = {
+    def mayNeedSignature(t: Type): Boolean = t match
+      case ExprType(e) => mayNeedSignature(e)
+      case MethodTpe(_, ps, res) => (!sym0.isConstructor && mayNeedSignature(res)) || ps.exists(mayNeedSignature)
+      case _: TypeRef => !t.isAny && !t.isAnyRef && !t.isRef(defn.StringClass, skipRefined = false) && !t.isPrimitiveValueType
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:43]: Not sure if there's a better way to ask "is this definitely not a generic type or parameter"

**Comment by SolalPirelli** [comment]
OK this is finally ready for a proper review, perf is basically the same as main (a little better or worse on some of the benchmarks but overall a wash) and it fixes a number of bugs while also simplifying logic

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala (hunk 9 above)
Comment (tanishiking) [reply compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:438]: While it seems it doesn't happen anymore with current implementation, what do you think about adding an assertion to prevent infinite recursion?
That would help us to spot, if we re-introduce the bug in future.

```scala
    val etp = erasure(tp)
    assert(etp ne tp, i"Cannot generate generic signature for $tp")
    jsig(etp, toplevel, vcBoxing)
```

File: compiler/src/dotty/tools/backend/jvm/BTypeLoader.scala (hunk 4 above)
Comment (tanishiking) [reply compiler/src/dotty/tools/backend/jvm/BTypeLoader.scala:297]: Nice catch, TIL 😄

File: tests/generic-java-signatures/derivedNames.scala (hunk 11 above)
Comment (tanishiking) [reply tests/generic-java-signatures/derivedNames.scala]: [maybe action required]
Not very sure about the usage, but having outer class's type parameter aligns with Java:

```java
// Outer.java
public class Outer<A> {
  public class Inner {}
  public Inner inner() {
    return null;
  }
}
```

`scala-cli compile --server=false --power --compile-output classes Outer.java`

`javap -v classes.Outer`

```java
  public Outer<A>.Inner inner();
    descriptor: ()LOuter$Inner;
    ...
    Signature: #14                          // ()LOuter<TA;>.Inner;
```

**Review by tanishiking — CHANGES_REQUESTED** [review]
Thanks @SolalPirelli a lot for implementing this with performance in mind! Overall change looks good to me.
One thing I'm concerned is the test behavior change in `tests/generic-java-signatures/derivedNames.scala`.

Dropping the prefix-based `classSig` loses owner type information for inner classes, and it differs from javac equivalent. It would be nice to keep align with javac in generic signature generation.

File: tests/generic-java-signatures/derivedNames.scala (hunk 11 above)
Comment (SolalPirelli) [reply tests/generic-java-signatures/derivedNames.scala]: indeed, I jumped to conclusions too quickly...Java is happy to do `Foo<T1>.A<Foo<T1>.B>`

Comment (SolalPirelli) [reply tests/generic-java-signatures/derivedNames.scala]: ...I wanted to show whether javac can deal with the signature anyway but it turns out even on main we emit nonsensical inner classes for this case, so let me fix that in another PR first 🙃

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala (hunk 9 above)
Comment (SolalPirelli) [reply compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:438]: good idea, done

File: tests/generic-java-signatures/derivedNames.scala (hunk 11 above)
Comment (SolalPirelli) [reply tests/generic-java-signatures/derivedNames.scala]: OK, so... short version, javac is fine with this change, as the test I added in #26137 still passes: https://github.com/scala/scala3/pull/26137/changes

Long version, `out2` makes sense to me as it is now, because we have the following inner class declarations in our output (along with the corresponding generic signatures):
```
Foo=class Test$Foo of class Test (sig: Test$Foo<T1>)
A=class Test$Foo$A of class Test$Foo (sig: Test$Foo$A<T2>)
B$=class Test$Foo$B$ of class Test$Foo (sig: Test$Foo$B$)
```

Java also does not include the generic parameters of the outer in the class's signature, i.e., it emits `Foo$A<T2>` for `class Foo<T1> { class A<T2>{ } }`

So it makes sense that `A[B.type]` would be `Test$Foo$A<Test$Foo$B$>`.... except that's not what javac does, since it indeed emits the generic parameters of the outer _when in a method signature_.

Also I have no clue what this `out1` expectation is or why it'd happen only sometimes on Windows.

Comment (tanishiking) [reply tests/generic-java-signatures/derivedNames.scala]: I was considering what might break with this change... one thing I found is that, the following test fails with this branch (while it pass in `main`, works in Scala2, and Java-equivalent test also works fine)

```scala
// Defs_1.scala
package genericouterinnermember

class Outer[A]:
  class Inner:
    def use(a: A): A = a
  def inner: Inner = new Inner
```

```java
// Use_2.java
package genericouterinnermember;

public class Use_2 {
    public static String use(Outer<String> outer) {
        return outer.inner().use("ok");
    }
}
```
```scala
// Test_3.scala
import genericouterinnermember.*

object Test:
  def main(args: Array[String]): Unit =
    val outer = new Outer[String]
    println(Use_2.use(outer))
```

fails with

  ```text
  tests/generic-java-signatures/generic-outer-inner-member/Use_2.java:5:
  error: incompatible types: Object cannot be converted to String
          return outer.inner().use("ok");
                                  ^
  ```

The issue is that Java no longer sees `outer.inner()` as an inner value tied to `Outer<String>` (because `inner()` no longer have a generic signature since it's same as descriptor), so `Inner#use(A): A` is seen as returning `Object` instead of `String`.

So, I think we should keep the  parameterized owner type in signature.

(and I also have no idea why signature is platform dependent ¯\_(ツ)_/¯)

**Review by tanishiking — CHANGES_REQUESTED** [review]
Thanks for fixing in upstream! I left a comment on `derivedNames.scala` test, now I think I found a concrete test case that regress from `main` and Scala2

https://github.com/scala/scala3/pull/25880#discussion_r3288739065

File: tests/generic-java-signatures/derivedNames.scala (hunk 11 above)
Comment (SolalPirelli) [reply tests/generic-java-signatures/derivedNames.scala]: Nice catch, I hadn't thought of this, thanks. I added this test + also a case in it with `@targetName`.

**Review by tanishiking — APPROVED** [review]
Thank you so much for addressing every details, LGTM :tada:
